src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -1476,6 +1476,22 @@ protected override void HandleStackTraceRequestAsync(IRequestResponder<StackTrac
                         });
                     }
 
+                    if (!IsClientVS)
+                    {
+                        Parallel.ForEach(response.StackFrames, frame =>
```

Comment:
my understanding is that Paralle.ForEach uses the default Task Schedular to spawn it's delegates, in which case we are spawning a task that spawns a task for each stack frame. Wouldn't we get the same level of parallelism by using a straight for loop to spawn tasks, collecting them, and rolling them up in a Task.WaitAll or something?

cc @gregg-miskelly

Reply:
Yes, that was my suggested fix here: https://github.com/microsoft/MIEngine/pull/1038#discussion_r479437530

Reply:
ah my bad, sorry.

Reply:
True, WaitAll should make things easier.
```cs
var frameTasks = response.StackFrames.Select(frame => Task.Run(() => {
    if (frame.Source == null || File.Exists(frame.Source.Path))
        return;

    // mark the stack frame to be skipped by default
    frame.Source.PresentationHint = Source.PresentationHintValue.Deemphasize;
})).ToArray();
try Task.WaitAll(frameTasks, 1000);
```

Reply:
@Trass3r this proposed change has the exact same issue @gregg-miskelly originally advocated against: if File.Exists hangs, one of your frameTasks will hang and Task.WaitAll will hang.

Reply:
Then I don't get what the [WaitAll](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitall#System_Threading_Tasks_Task_WaitAll_System_Threading_Tasks_Task___System_Int32_) overload with timeout is for.

And wouldn't the await whenall solution leave zombie tasks behind in case of hangs?

Reply:
Ah my apologies, I didn't see a the timeout in `WaitAll`. That timeout will work as you expect. Sorry again.

One potential issue still exists here: if the the timeout is triggered, the code will move past the `Task.WaitAll`, but the delegate will still be running and may complete later and try to set `frame.Source.PresentationHint` at a unknown time in the futue.

A fix for this can be seen in gregg-miskelly's initial suggestion which cancels a `CancellationTokenSource` after getting past the `Task.WaitAll`. The delegate is gated by the cancellation and doesn't try to set `frame.Source.PresentationHint` if the TokenSource was canceled.